pkg/spiffe/spiffe.go

```diff
@@ -221,3 +230,80 @@ func RetrieveSpiffeBundleRootCerts(config map[string]string, extraTrustedCerts [
 	}
 	return ret, nil
 }
+
+// PeerCertVerifier is an instance to verify the peer certificate in the SPIFFE way using the retrieved root certificates.
+type PeerCertVerifier struct {
+	generalCertPool *x509.CertPool
+	certPools       map[string]*x509.CertPool
+}
+
+// NewPeerCertVerifier returns a new PeerCertVerifier.
+func NewPeerCertVerifier() *PeerCertVerifier {
+	return &PeerCertVerifier{
+		generalCertPool: x509.NewCertPool(),
+		certPools:       make(map[string]*x509.CertPool),
+	}
+}
+
+// GetGeneralCertPool returns generalCertPool containing all root certs.
+func (v *PeerCertVerifier) GetGeneralCertPool() *x509.CertPool {
+	return v.generalCertPool
+}
+
+// AddMapping adds a new trust domain to certificates mapping to the certPools map.
+func (v *PeerCertVerifier) AddMapping(trustDomain string, certs []*x509.Certificate) {
+	if v.certPools[trustDomain] == nil {
+		v.certPools[trustDomain] = x509.NewCertPool()
+	}
+	for _, cert := range certs {
+		v.certPools[trustDomain].AddCert(cert)
+		v.generalCertPool.AddCert(cert)
+	}
+	spiffeLog.Infof("Added %d certs to trust domain %s in peer cert verifier", len(certs), trustDomain)
+}
+
+// AddMappings merges a trust domain to certs map to the certPools map.
+func (v *PeerCertVerifier) AddMappings(certMap map[string][]*x509.Certificate) {
+	for trustDomain, certs := range certMap {
+		v.AddMapping(trustDomain, certs)
+	}
+}
+
+// VerifyPeerCert is an implementation of tls.Config.VerifyPeerCertificate.
+// It verifies the peer certificate using the root certificates associated with its trust domain.
+func (v *PeerCertVerifier) VerifyPeerCert(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+	if len(rawCerts) == 0 {
+		// Peer doesn't present a certificate. Just skip. Other authn methods may be used.
+		return nil
+	}
+	var peerCert *x509.Certificate
+	intCertPool := x509.NewCertPool()
+	for id, rawCert := range rawCerts {
+		cert, err := x509.ParseCertificate(rawCert)
```

Comment:
Shall we use ParseCertificates https://golang.org/pkg/crypto/x509/#ParseCertificates?

Reply:
This is a single cert. I think using ParseCertificate should be fine.